tensorboard/components/tensor_widget/menu.ts

```diff
@@ -0,0 +1,278 @@
+  options: string[];
+
+  /** The default selection from `options`, a 0-based index. */
+  defaultSelection: number;
+
+  /**
+   * The callback that gets called when the selection has changed.
+   *
+   * The `currentSelection` argument is the 0-based index for the currently
+   * selected option.
+   */
+  callback: (currentSelection: number) => void | Promise<void>;
+}
+
+export interface MenuConfig {
+  /** An ordered list of items that comprise the menu. */
+  items: MenuItemConfig[];
+}
+
+/**
+ * The configuration of an item of a FlatMenu.
+ */
+interface FlatMenuItemConfig {
+  /** Caption of the FlatMenu item. */
+  caption: string;
+
+  /** The click callback for the item. */
+  onClick: EventCallback | null;
+
+  /** The hover callback for the item. */
+  onHover: EventCallback | null;
+
+  /**
+   * Whether the item is disabled.
+   *
+   * If not specified, treated as `false` (not disabled).
+   */
+  disabled?: boolean;
+}
+
+/**
+ * Helper class: A menu item without hierarchy.
+ *
+ * A FlatMenu doesn't support hierarchy. The `Menu` class below, which supports
+ * menus with hierarchy, is built by composing this helper class.
+ */
+class FlatMenu {
+  private isShown: boolean = false;
+  private dropdown: HTMLDivElement;
+
+  /**
+   * Constructor of FlatMenu.
+   * @param parentElement The parent element that the root UI element of this
+   *   FlatMenu will be appended to as a child.
+   */
+  constructor(parentElement: HTMLDivElement) {
+    this.dropdown = document.createElement('div');
+    this.dropdown.classList.add('tensor-widget-dim-dropdown');
+    this.dropdown.style.position = 'fixed';
+    this.dropdown.style.display = 'none';
+    this.dropdown.addEventListener('mouseleave', () => {
+      this.hide();
+    });
+    parentElement.appendChild(this.dropdown);
+  }
+
+  /**
+   * Show the FlatMenu item.
+   * @param top The top coordinate for the FlatMenu.
+   * @param left The left coordinate for the FlatMenu.
+   * @param itemConfigs Configuration of the menu items.
+   */
+  show(top: number, left: number, itemConfigs: FlatMenuItemConfig[]) {
+    itemConfigs.forEach((itemConfig, i) => {
+      const menuItem = document.createElement('div');
+      menuItem.classList.add('tensor-widget-dim-dropdown-menu-item');
+      menuItem.textContent = itemConfig.caption;
+      this.dropdown.appendChild(menuItem);
+      if (itemConfig.disabled) {
+        menuItem.classList.add('tensor-widget-dim-dropdown-menu-item-disabled');
+        return;
+      }
+
+      menuItem.addEventListener('click', (event) => {
+        if (itemConfig.onClick !== null) {
+          itemConfig.onClick(event);
+        }
+        this.hide();
+      });
+      menuItem.addEventListener('mouseenter', (event) => {
+        if (itemConfig.onHover !== null) {
+          itemConfig.onHover(event);
+        }
+        menuItem.classList.add('tensor-widget-dim-dropdown-menu-item-active');
+      });
+      menuItem.addEventListener('mouseleave', () => {
```

Comment:
Ideally mouseleave should be a no-op, and this logic should trigger on a click anywhere outside the menu.  Maybe that requires a transparent page-sized element to catch the clicks?  I don't know how Angular Material components handle this case.  I did a React version previously; I don't remember how it worked but let me know if it would help to look it up.

Reply:
Done. Made mouseleave a no-op as you suggested.